pkg/services/ngalert/schedule/schedule.go

```diff
@@ -139,21 +152,162 @@ func (sch *schedule) Unpause() error {
 }
 
 func (sch *schedule) Run(ctx context.Context) error {
-	sch.wg.Add(1)
+	var wg sync.WaitGroup
+	wg.Add(2)
 
 	go func() {
+		defer wg.Done()
 		if err := sch.ruleEvaluationLoop(ctx); err != nil {
 			sch.log.Error("failure while running the rule evaluation loop", "err", err)
 		}
 	}()
 
-	sch.wg.Wait()
+	go func() {
+		defer wg.Done()
+		if err := sch.adminConfigSync(ctx); err != nil {
+			sch.log.Error("failure while running the admin configuration sync", "err", err)
+		}
+	}()
+
+	wg.Wait()
 	return nil
 }
 
-func (sch *schedule) ruleEvaluationLoop(ctx context.Context) error {
-	defer sch.wg.Done()
+// SyncAndApplyConfigFromDatabase looks for the admin configuration in the database and adjusts the sender(s) accordingly.
+func (sch *schedule) SyncAndApplyConfigFromDatabase() error {
+	sch.log.Debug("start of admin configuration sync")
+	cfgs, err := sch.adminConfigStore.GetAdminConfigurations()
+	if err != nil {
+		return err
+	}
+
+	sch.log.Debug("found admin configurations", "count", len(cfgs))
+
+	orgsFound := make(map[int64]struct{}, len(cfgs))
+	sch.sendersMtx.Lock()
+	for _, cfg := range cfgs {
+		orgsFound[cfg.OrgID] = struct{}{} // keep track of the which senders we need to keep.
+
+		existing, ok := sch.senders[cfg.OrgID]
+
+		// If the tenant has no Alertmanager(s) configured and no running sender no-op.
+		if !ok && len(cfg.Alertmanagers) == 0 {
+			sch.log.Debug("no external alertmanagers configured", "org", cfg.OrgID)
+			continue
+		}
+
+		// We have a running sender but no Alertmanager(s) configured, shut it down.
+		if ok && len(cfg.Alertmanagers) == 0 {
+			sch.log.Debug("no external alertmanager(s) configured, sender will be stopped", "org", cfg.OrgID)
+			delete(orgsFound, cfg.OrgID)
+			continue
+		}
+
+		// We have a running sender, check if we need to apply a new config.
+		if ok {
+			if sch.sendersCfgHash[cfg.OrgID] == cfg.AsSHA256() {
+				sch.log.Debug("sender configuration is the same as the one running, no-op", "org", cfg.OrgID, "alertmanagers", cfg.Alertmanagers)
+				continue
+			}
+
+			sch.log.Debug("applying new configuration to sender", "org", cfg.OrgID, "alertmanagers", cfg.Alertmanagers)
+			err := existing.ApplyConfig(cfg)
+			if err != nil {
+				sch.log.Error("failed to apply configuration", "err", err, "org", cfg.OrgID)
+			}
+			sch.sendersCfgHash[cfg.OrgID] = cfg.AsSHA256()
```

Comment:
We should update this only when `err == nil`, else it won't retry to apply config. If we don't want to retry, then ignore this comment :)

Reply:
My bad, I think I got rid of the `continue` by accident that.

Reply:
I've added more tests cases that cover this.